File: app/src/main/java/com/aracroproducts/attention/MainActivity.java
package com.aracroproducts.attention;

import android.app.AlertDialog;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.VibrationEffect;
import android.os.Vibrator;
import android.provider.Settings;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.localbroadcastmanager.content.LocalBroadcastManager;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.GoogleApiAvailability;
import com.google.android.gms.tasks.Task;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.iid.FirebaseInstanceId;
import com.google.firebase.iid.InstanceIdResult;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.security.NoSuchAlgorithmException;
import java.security.spec.InvalidKeySpecException;
import java.util.HashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

import javax.crypto.SecretKey;
import javax.crypto.SecretKeyFactory;
import javax.crypto.spec.PBEKeySpec;

public class MainActivity extends AppCompatActivity {

    public static final String USER_INFO = "user";
    public static final String FRIENDS = "listen";

    public static final String MY_ID = "id";
    public static final String MY_NAME = "name";
    public static final String MY_TOKEN = "token";
    public static final String UPLOADED = "uploaded";
    public static final String FRIEND_LIST = "friends";
    public static final String OVERLAY_NO_PROMPT = "OverlayDoNotAsk";

    private static final int COMPAT_HEAVY_CLICK = 5;

    public static final int NAME_CALLBACK = 0;
    public static final int EDIT_NAME_CALLBACK = 1;

    private final String TAG = getClass().getName();

    private String token;
    private User user;

    private final BroadcastReceiver networkCallback = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            Log.d(TAG, "Received callback from network class");
            SharedPreferences.Editor editor = getSharedPreferences(USER_INFO, Context.MODE_PRIVATE).edit();
            int resultCode = intent.getIntExtra(AppServer.EXTRA_RESULT_CODE, AppServer.CODE_NA);

            switch (Objects.requireNonNull(intent.getAction())) {
                case AppServer.ACTION_POST_TOKEN:
                    if (resultCode == AppServer.CODE_SUCCESS) {
                        editor.putBoolean(UPLOADED, true);
                        editor.putString(MY_TOKEN, token);
                        Toast.makeText(MainActivity.this, getString(R.string.user_registered), Toast.LENGTH_SHORT).show();
                    } else {
                        editor.putBoolean(UPLOADED, false);
                    }
                    break;
                case AppServer.ACTION_SEND_ALERT:
                    if (resultCode == AppServer.CODE_SUCCESS) {
                        View layout = findViewById(R.id.coordinatorLayout);
                        Snackbar snackbar = Snackbar.make(layout, R.string.alert_sent, Snackbar.LENGTH_SHORT);
                        snackbar.show();
                    } else {
                        Toast.makeText(MainActivity.this, getString(R.string.general_error), Toast.LENGTH_LONG).show();
                    }
            }
            editor.apply();
        }
    };

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(this) != ConnectionResult.SUCCESS) {
            Toast.makeText(this, getString(R.string.no_play_services), Toast.LENGTH_LONG).show();
            GoogleApiAvailability.getInstance().makeGooglePlayServicesAvailable(this);
            return;
        }

        SharedPreferences prefs = getSharedPreferences(USER_INFO, Context.MODE_PRIVATE);
        SharedPreferences settings = PreferenceManager.getDefaultSharedPreferences(this);

        String[] notificationValues = getResources().getStringArray(R.array.notification_values);

        if (!settings.contains(getString(R.string.ring_preference_key))) {
            SharedPreferences.Editor settingsEditor = settings.edit();
            Set<String> ringAllowed = new HashSet<>();
            ringAllowed.add(notificationValues[2]);
            settingsEditor.putStringSet(getString(R.string.ring_preference_key), ringAllowed);
            settingsEditor.apply();
        }

        if (!settings.contains(getString(R.string.vibrate_preference_key))) {
            SharedPreferences.Editor settingsEditor = settings.edit();
            Set<String> vibrateAllowed = new HashSet<>();
            vibrateAllowed.add(notificationValues[1]);
            vibrateAllowed.add(notificationValues[2]);
            settingsEditor.putStringSet(getString(R.string.vibrate_preference_key), vibrateAllowed);
            settingsEditor.apply();
        }

        if (!prefs.contains(UPLOADED)) {
            SharedPreferences.Editor editor = prefs.edit();
            editor.putBoolean(UPLOADED, false);
            editor.apply();
        }

        token = prefs.getString(MY_TOKEN, null);

        if (!settings.contains(getString(R.string.name_key)) || !prefs.contains(MY_ID)) {
            user = new User();
            Intent intent = new Intent(this, DialogActivity.class);
            startActivityForResult(intent, NAME_CALLBACK);
        } else {
            user = new User(prefs.getString(MY_ID, null), token);
        }

        if (user.getUid() != null && user.getToken() != null && !prefs.getBoolean(UPLOADED, false)) {
            updateToken(user);
        } else {
            getToken();
        }

        FloatingActionButton fab = findViewById(R.id.fab);
        fab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Log.d(this.getClass().getName(), "Attempting to add");
                Intent intent = new Intent(view.getContext(), Add.class);
                startActivity(intent);
            }
        });

        if (!Settings.canDrawOverlays(this) && !prefs.contains(OVERLAY_NO_PROMPT)) {
            androidx.appcompat.app.AlertDialog alertDialog = new androidx.appcompat.app.AlertDialog.Builder(this).create();
            alertDialog.setTitle(getString(R.string.draw_title));
            alertDialog.setMessage(getString(R.string.draw_message));
            alertDialog.setButton(DialogInterface.BUTTON_POSITIVE, getString(R.string.open_settings), (dialogInterface, i) -> {
                Intent intent = new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:" + getApplicationContext().getPackageName()));
                startActivity(intent);
            });
            alertDialog.setButton(DialogInterface.BUTTON_NEGATIVE, getString(R.string.do_not_ask_again), (dialogInterface, i) -> {
                SharedPreferences.Editor editor = prefs.edit();
                editor.putBoolean(OVERLAY_NO_PROMPT, true);
                editor.apply();
            });
            alertDialog.show();
        }


    }

    /**
     * Receives results from input dialog boxes that are displayed to the user
     * @param requestCode   - The code corresponding to the type of request that was made
     * @param resultCode    - The result of the dialog (not used)
     * @param data          - Contains the user input
     */
    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (requestCode) {
            case NAME_CALLBACK: // when the user adds or changes their name
                SharedPreferences prefs = getSharedPreferences(USER_INFO, Context.MODE_PRIVATE);
                SharedPreferences.Editor editor = prefs.edit();
                SharedPreferences.Editor settingsEditor = PreferenceManager.getDefaultSharedPreferences(this).edit();

                settingsEditor.putString(getString(R.string.name_key), data.getStringExtra(MY_NAME));
                editor.putString(MY_ID, makeId(data.getStringExtra(MY_NAME)));

                editor.apply();
                settingsEditor.apply();

                user.setUid(prefs.getString(MY_ID, null));
                addUserToDB(user);
                break;
            case EDIT_NAME_CALLBACK: // when the user changes one of their friends' names
                Log.d(TAG, "Received edit name callback");
                SharedPreferences friends = getSharedPreferences(FRIENDS, Context.MODE_PRIVATE);
                List<String[]> friendList = parseFriends(friends.getString(FRIEND_LIST, null));

                if (friendList == null) {
                    Log.w(TAG, "FriendList was null, unable to edit");
                    break;
                }

                String friendId = data.getStringExtra(DialogActivity.EXTRA_USER_ID);

                String[] friend = null;
                int friendIndex = -1;

                for (int i = 0; i < friendList.size(); i++) {
                    if (friendList.get(i)[1].equals(friendId)) {
                        friend = friendList.get(i);
                        friendIndex = i;
                        break;
                    }
                }

                if (friend == null) {
                    Log.e(TAG, "Could not find requested ID to rename");
                    break;
                }

                friend[0] = data.getStringExtra(MY_NAME);

                friendList.set(friendIndex, friend);

                Gson gson = new Gson();
                String friendJson = gson.toJson(friendList);

                SharedPreferences.Editor friendEditor = friends.edit();
                friendEditor.putString(FRIEND_LIST, friendJson);
                friendEditor.apply();
                populateFriendList();

        }
        super.onActivityResult(requestCode, resultCode, data);
    }

    /**
     * Helper method that makes a unique user ID
     * @param name  - The name that the ID is based on
     * @return      - The ID
     */
    private String makeId(String name) {
        String fullString = name == null? "" : name + Build.FINGERPRINT;
        byte[] salt = {69, 42, 0, 37, 10, 127, 34, 85, 83, 24, 98, 75, 49, 8, 67}; // very secure salt but this isn't a cryptographic application so it doesn't really matter

        try {
            SecretKeyFactory secretKeyFactory;
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) secretKeyFactory = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA512"); //not available to Android 7.1 and lower
            else secretKeyFactory = SecretKeyFactory.getInstance("PBKDF2withHmacSHA1"); // available since API 10
            PBEKeySpec spec = new PBEKeySpec(fullString.toCharArray(), salt, 32, 64);
            SecretKey key = secretKeyFactory.generateSecret(spec);
            byte[] hashed = key.getEncoded();
            StringBuilder builder = new StringBuilder();
            for (byte letter : hashed) {
                builder.append(letter);
            }
            return builder.toString();
        } catch (InvalidKeySpecException | NoSuchAlgorithmException e) {
            e.printStackTrace();
            throw new RuntimeException();
        }
    }

    /**
     * Helper method that gets the Firebase token
     */
    private void getToken() {
        FirebaseInstanceId.getInstance().getInstanceId().addOnCompleteListener(task -> {
            if (!task.isSuccessful()) {
                Log.w(TAG, "getInstanceId failed", task.getException());
                return;
            }

            // Get new Instance ID token
            token = task.getResult().getToken();
            Log.d(TAG, "Got token! " + token);
            user.setToken(token);

            SharedPreferences preferences = getSharedPreferences(USER_INFO, Context.MODE_PRIVATE);
            if (!token.equals(preferences.getString(MY_TOKEN, ""))) {
                if (user.getUid() == null && preferences.getBoolean(UPLOADED, false)) {
                    preferences.edit().putBoolean(UPLOADED, false).apply();
                } else {
                    updateToken(user);
                }
            }

            // Log and toast
            String msg = getString(R.string.msg_token_fmt, token);
            Log.d(TAG, msg);
        });
    }

    @Override
    public void onResume() {
        super.onResume();

        if (GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(this) != ConnectionResult.SUCCESS) {
            Toast.makeText(this, getString(R.string.no_play_services), Toast.LENGTH_LONG).show();
            GoogleApiAvailability.getInstance().makeGooglePlayServicesAvailable(this);
            return;
        }

        populateFriendList();

        IntentFilter filter = new IntentFilter(AppServer.ACTION_SEND_ALERT);
        filter.addAction(AppServer.ACTION_POST_TOKEN);
        LocalBroadcastManager manager = LocalBroadcastManager.getInstance(this);
        manager.registerReceiver(networkCallback, filter);

    }

    @Override
    public void onPause() {
        super.onPause();
        LocalBroadcastManager manager = LocalBroadcastManager.getInstance(this);
        manager.unregisterReceiver(networkCallback);
    }

    /**
     * Helper method to put all the friends into the recycler view
     */
    private void populateFriendList() {
        SharedPreferences friends = getSharedPreferences(FRIENDS, Context.MODE_PRIVATE);
        String friendJson = friends.getString(FRIEND_LIST, null);

        List<String[]> friendList = parseFriends(friendJson);

        if (friendList == null) return;

        RecyclerView friendListView = findViewById(R.id.friends_list);
        friendListView.setLayoutManager(new LinearLayoutManager(this));

        String[][] dataset = new String[friendList.size()][2];
        for (int x = 0; x < friendList.size(); x++) {
            dataset[x][0] = friendList.get(x)[0];
            dataset[x][1] = friendList.get(x)[1];
        }

        FriendAdapter adapter = new FriendAdapter(dataset, null);

        FriendAdapter.Callback adapterListener = new FriendAdapter.Callback() {
            @Override
            public void onSendAlert(String id, String message) {
                sendAlertToServer(id, message);
            }

            @Override
            public void onDeletePrompt(int position, String name) {
                AlertDialog alertDialog = new AlertDialog.Builder(MainActivity.this).create();
                alertDialog.setTitle(getString(R.string.confirm_delete_title));
                alertDialog.setMessage(getString(R.string.confirm_delete_message, name));
                alertDialog.setButton(DialogInterface.BUTTON_POSITIVE, getString(R.string.yes), (dialogInterface, i) -> {
                    deleteFriend(position);
                    dialogInterface.cancel();
                });
                alertDialog.setButton(DialogInterface.BUTTON_NEGATIVE, getString(android.R.string.cancel), (dialogInterface, i) -> dialogInterface.cancel());
                alertDialog.show();
            }

            @Override
            public void onEditName(String id) {
                Intent intent = new Intent(MainActivity.this, DialogActivity.class);
                intent.putExtra(DialogActivity.EXTRA_EDIT_NAME, true);
                intent.putExtra(DialogActivity.EXTRA_USER_ID, id);
                startActivityForResult(intent, EDIT_NAME_CALLBACK);
            }

            @Override
            public void onLongPress() {
                Vibrator vibrator = (Vibrator) getSystemService(Context.VIBRATOR_SERVICE);
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                    VibrationEffect effect;
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) effect = VibrationEffect.createPredefined(VibrationEffect.EFFECT_HEAVY_CLICK);
                    else effect = VibrationEffect.createOneShot(100, COMPAT_HEAVY_CLICK);
                    vibrator.vibrate(effect);
                } else {
                    vibrator.vibrate(100);
                }
            }

        };

        adapter.setCallback(adapterListener);

        friendListView.setAdapter(adapter);
    }

    /**
     * Converts a json string of friends into a List of {friend name, friend ID} pairs
     * @param json  - JSON string to parse
     * @return      - The list of friends
     */
    public static List<String[]> parseFriends(String json) {
        if (json == null) return null;
        Gson gson = new Gson();

            Type arrayListType = new TypeToken<List<String[]>>() {
            }.getType();
            return gson.fromJson(json, arrayListType);
    }

    /**
     * Deletes the user's friend with the given index
     * @param index - The index that the friend is at
     */
    private void deleteFriend(int index) {
        SharedPreferences friends = getSharedPreferences(FRIENDS, Context.MODE_PRIVATE);
        String friendJson = friends.getString(FRIEND_LIST, null);
        List<String[]> friendList;


        if (friendJson == null) {
            Log.w(TAG, "Friend list was null, unable to delete friend");
            return;
        }

        Gson gson = new Gson();
        friendList = parseFriends(friendJson);

        friendList.remove(index);
        Log.d(TAG, "Removed friend");
        SharedPreferences.Editor editor = friends.edit();
        editor.putString(FRIEND_LIST, gson.toJson(friendList));
        editor.apply();
        populateFriendList();

    }

    /**
     * Sends the user's notification to the server
     * @param id        - The recipient ID
     * @param message   - The message to send to the person
     */
    private void sendAlertToServer(String id, String message) {
        Log.d(TAG, "Sending alert to server via AppServer service");
        //PendingIntent pendingIntent = createPendingResult(AppServer.CALLBACK_POST_TOKEN, new Intent(), 0);
        Intent intent = new Intent(MainActivity.this, AppServer.class);


        intent.putExtra(AppServer.EXTRA_TO, id);
        intent.putExtra(AppServer.EXTRA_FROM, user.getUid());
        if (message != null) {
            intent.putExtra(AppServer.EXTRA_MESSAGE, message);
        }
        intent.setAction(AppServer.ACTION_SEND_ALERT);
       // intent.putExtra(AppServer.EXTRA_PENDING_RESULT, pendingIntent);
        AppServer.enqueueWork(this, intent);


    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        if (id == R.id.action_settings) {
            Intent intent = new Intent(this, SettingsActivity.class);
            startActivity(intent);
        }

        return super.onOptionsItemSelected(item);
    }


    /**
     * Adds the specified new user to the Aracro Products database
     * @param user  - The user to add to the database
     */
    private void addUserToDB(User user) {

        Task<InstanceIdResult> idResultTask = FirebaseInstanceId.getInstance().getInstanceId();
        idResultTask.addOnCompleteListener(task -> {
            user.setToken(task.getResult().getToken());

            //PendingIntent pendingIntent = createPendingResult(AppServer.CALLBACK_POST_TOKEN, new Intent(), 0);
            Intent intent = new Intent(MainActivity.this, AppServer.class);

            intent.putExtra(AppServer.EXTRA_TOKEN, user.getToken());
            intent.putExtra(AppServer.EXTRA_ID, user.getUid());
            intent.setAction(AppServer.ACTION_POST_TOKEN);
            //intent.putExtra(AppServer.EXTRA_PENDING_RESULT, pendingIntent);
            AppServer.enqueueWork(MainActivity.this, intent);
            Log.d(TAG, getString(R.string.log_sending_msg));
        });
    }

    /**
     * Alias for addUserToDB()
     * @param user  - the user to add to the database
     */
    private void updateToken(User user) {
        addUserToDB(user);
        /*DatabaseReference database = FirebaseDatabase.getInstance().getReference();
        if (user.getUid() != null) {
            database.child("users").child(user.getUid()).child("token").setValue(user.getToken());
        }*/
    }

}

File: app/src/main/java/com/aracroproducts/attention/MainActivity.kt
package com.aracroproducts.attention

import android.content.*
import android.net.Uri
import android.os.*
import android.provider.Settings
import android.util.Log
import android.widget.Toast
import androidx.activity.compose.setContent
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.*
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material.*
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Add
import androidx.compose.material.icons.filled.Close
import androidx.compose.material.icons.filled.Settings
import androidx.compose.runtime.*
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.blur
import androidx.compose.ui.text.input.KeyboardCapitalization
import androidx.compose.ui.text.input.KeyboardType
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.preference.PreferenceManager
import com.aracroproducts.attention.MainViewModel.Companion.MY_NAME
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {
    private val sTAG = javaClass.name
    private val friendModel: MainViewModel by viewModels()

    enum class State {
        NORMAL, CONFIRM, CANCEL, EDIT
    }

    /**
     * Called when the activity is created
     *
     * @param savedInstanceState    Instance data saved from before the activity was killed
     */
    @OptIn(ExperimentalFoundationApi::class)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        // Creates a notification channel for displaying failed-to-send notifications
        MainViewModel.createFailedAlertNotificationChannel(this)

        if (!checkPlayServices()) return

        friendModel.loadUserPrefs()

        setContent {
            MaterialTheme(colors = if (isSystemInDarkTheme()) darkColors else lightColors) {
                HomeWrapper(model = friendModel)
            }
        }
    }

    @ExperimentalFoundationApi
    @Composable
    fun HomeWrapper(model: MainViewModel) {
        val displayDialog: Triple<MainViewModel.DialogStatus, Friend?, (String) -> Unit> by model
                .dialogState
        val showSnackbar = model.isSnackBarShowing

        Home(
                friends = model.friends.value ?: listOf(),
                onLongPress = { model.onLongPress() },
                onEditName = { model.onEditName(it) },
                onDeletePrompt = { model.onDeleteFriend(it) },
                dialogState = displayDialog,
                showSnackbar = showSnackbar
        )
    }

    @ExperimentalFoundationApi
    @Composable
    fun Home(friends: List<Friend>, onLongPress: () -> Unit, onEditName: (friend: Friend) -> Unit,
             onDeletePrompt: (friend: Friend) -> Unit,
             dialogState: Triple<MainViewModel.DialogStatus, Friend?, (String) -> Unit>,
             showSnackbar:
             Boolean) {


        val scaffoldState = rememberScaffoldState()
        val scope = rememberCoroutineScope()
        if (showSnackbar) {
            LaunchedEffect(scaffoldState.snackbarHostState) {
                scope.launch {
                    val result = scaffoldState.snackbarHostState.showSnackbar(message = getString(R
                            .string
                            .alert_sent))  // cancels by default after a short amount of time

                    when (result) {
                        SnackbarResult.Dismissed -> {
                            friendModel.dismissSnackBar()
                        }
                        else -> {}
                    }
                }
            }
        }
        when (dialogState.first) {
            MainViewModel.DialogStatus.USERNAME -> UserNameDialog()
            MainViewModel.DialogStatus.OVERLAY_PERMISSION -> OverlaySettingsDialog()
            MainViewModel.DialogStatus.ADD_MESSAGE_TEXT -> dialogState.second?.let {
                AddMessageText(friend = it, onSend = dialogState.third) }
            MainViewModel.DialogStatus.FRIEND_NAME -> dialogState.second?.let {EditFriendNameDialog(
                    friend = it)}
            MainViewModel.DialogStatus.CONFIRM_DELETE -> dialogState.second?.let {DeleteFriendDialog(
                    friend = it)}
            MainViewModel.DialogStatus.NONE -> {}
        }

        Scaffold(scaffoldState = scaffoldState,
                topBar = {
                    TopAppBar(
                            backgroundColor = MaterialTheme.colors.primary,
                            title = { Text(getString(R.string.app_name)) },
                            actions = {
                                IconButton(onClick = {
                                    val intent = Intent(applicationContext,
                                            SettingsActivity::class.java)
                                    startActivity(intent)
                                }) {
                                    Icon(Icons.Filled.Settings, contentDescription = getString(R
                                            .string.action_settings))

                                }
                            }
                    )
                },
                floatingActionButton = {
                    FloatingActionButton(onClick = {
                        Log.d(this.javaClass.name, "Attempting to add")
                        val intent = Intent(this, Add::class.java)
                        startActivity(intent)
                    },
                            backgroundColor = MaterialTheme.colors.secondary) {
                        Icon(Icons.Filled.Add, contentDescription = getString(R.string.add_friend))
                    }
                }
        ) {
            LazyColumn(Modifier.background(MaterialTheme.colors.background)) {
                items(friends) { friend ->
                    FriendCard(
                            friend = friend,
                            onLongPress = onLongPress,
                            onEditName = onEditName,
                            onDeletePrompt = onDeletePrompt
                    )
                }
            }
        }
    }

    @Composable
    fun AddMessageText(friend: Friend, onSend: (message: String) -> Unit) {
        // on ok call onSend(textfield value)
        var message by rememberSaveable {
            mutableStateOf("")
        }
        AlertDialog(onDismissRequest = { friendModel.popDialogState() },
                confirmButton = {
                    TextButton(onClick = {
                            friendModel.popDialogState()
                        onSend(message)
                    }) {
                        Text( text = getString(R.string.save))
                    }
                },
                dismissButton = {
                    TextButton(onClick = {
                        friendModel.popDialogState()
                    }) {
                        Text(text = getString(R.string.cancel))
                    }
                },
                title = {Text(text = getString(R.string.rename))},
                text = { OutlinedTextField(
                        value = message,
                        onValueChange = { message = it },
                        keyboardOptions = KeyboardOptions(
                                keyboardType = KeyboardType.Text,
                                capitalization = KeyboardCapitalization.Words),
                        singleLine = false,
                        label = { Text(text = getString(R.string.message_label, friend.name)) },
                        placeholder = { Text(text = getString(R.string.message_hint)) }
                )}
        )
    }

    @Composable
    fun DeleteFriendDialog(friend: Friend) {
        AlertDialog(onDismissRequest = { friendModel.popDialogState() },
                confirmButton = {
                    TextButton(onClick = {
                        friendModel.popDialogState()
                        friendModel.confirmDeleteFriend(friend = friend)
                    }) {
                        Text(text = getString(R.string.delete))
                    }
                },
                dismissButton = {
                    TextButton(onClick = {
                        friendModel.popDialogState()
                    }) {
                        Text(text = getString(R.string.do_not_ask_again))
                    }
                },
                title = {Text(text = getString(R.string.confirm_delete_title))},
                text = { Text(text = getString(R.string.confirm_delete_message))}
        )
    }

    @Composable
    fun EditFriendNameDialog(friend: Friend) {
        var name by rememberSaveable {
            mutableStateOf("")
        }
        var error by remember { mutableStateOf(false) }
        AlertDialog(onDismissRequest = { friendModel.popDialogState() },
                confirmButton = {
                    TextButton(onClick = {
                        val savingName = name.trim()
                        if (savingName.isEmpty()) {
                            error = true
                        } else {
                            friendModel.confirmEditName(friend.id, savingName)
                            friendModel.popDialogState()
                        }
                    }) {
                        Text( text = getString(R.string.save))
                    }
                },
                dismissButton = {
                    TextButton(onClick = {
                        friendModel.popDialogState()
                    }) {
                        Text(text = getString(R.string.cancel))
                    }
                },
                title = {Text(text = getString(R.string.rename))},
                text = { OutlinedTextField(
                        value = name,
                        onValueChange = { name = it },
                        keyboardOptions = KeyboardOptions(
                                keyboardType = KeyboardType.Text,
                                capitalization = KeyboardCapitalization.Words),
                        singleLine = true,
                        label = { Text(text = getString(R.string.name)) },
                        isError = error,
                        placeholder = { Text(text = getString(R.string.new_name)) }
                )}
        )
    }

    @Composable
    fun OverlaySettingsDialog() {
        AlertDialog(onDismissRequest = { friendModel.popDialogState() },
                confirmButton = {
                    TextButton(onClick = {
                        val intent = Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                                Uri.parse("package:" + applicationContext.packageName))
                        friendModel.popDialogState()
                        startActivity(intent)
                    }) {
                        Text(text = getString(R.string.open_settings))
                    }
                },
                dismissButton = {
                    TextButton(onClick = { val editor = getSharedPreferences(MainViewModel.USER_INFO, MODE_PRIVATE).edit()
                        editor.putBoolean(MainViewModel.OVERLAY_NO_PROMPT, true)
                        editor.apply()
                        friendModel.popDialogState()
                    }) {
                        Text(text = getString(R.string.do_not_ask_again))
                    }
                },
                title = {Text(text = getString(R.string.draw_title))},
                text = { Text(text = getString(R.string.draw_message))}
                )
    }

    @Composable
    fun UserNameDialog() {
        var name by rememberSaveable {
            mutableStateOf("")
        }
        var error by remember { mutableStateOf(false) }
        AlertDialog(onDismissRequest = {},
                buttons = {
                    Row(
                            modifier = Modifier.padding(all = 8.dp),
                            horizontalArrangement = Arrangement.Center
                    ) {
                        TextButton(onClick = {
                            val savingName = name.trim()
                            if (savingName.isEmpty()) {
                                error = true
                            } else {
                                val editor = PreferenceManager
                                        .getDefaultSharedPreferences(this@MainActivity)
                                        .edit()
                                editor.putString(MY_NAME, savingName)
                                editor.apply()
                                friendModel.popDialogState()
                            }
                        },
                                modifier = Modifier.fillMaxWidth()) {
                            Text(getString(android.R.string.ok))
                        }
                    }
                },
                title = { Text(text = getString(R.string.name_dialog_title)) },
                text = {
                    OutlinedTextField(
                            value = name,
                            onValueChange = { name = it },
                            keyboardOptions = KeyboardOptions(
                                    keyboardType = KeyboardType.Text,
                                    capitalization = KeyboardCapitalization.Words),
                            singleLine = true,
                            label = { Text(text = getString(R.string.name)) },
                            isError = error,
                            placeholder = { Text(text = getString(R.string.placeholder_name)) }
                    )
                })
    }

    @ExperimentalFoundationApi
    @Composable
    fun FriendCard(friend: Friend, onLongPress: () -> Unit, onEditName: (friend: Friend) -> Unit,
                   onDeletePrompt: (friend: Friend) -> Unit) {
        var state by remember { mutableStateOf(State.NORMAL) }
        var message: String? by remember { mutableStateOf(null) }

        Box(modifier = Modifier
                .fillMaxWidth(1F)
                .padding(10.dp)
                .combinedClickable(onClick = {
                    state = when (state) {
                        State.NORMAL -> State.CONFIRM
                        State.CONFIRM, State.CANCEL, State.EDIT -> State.NORMAL
                    }
                }, onLongClick = {
                    state = when (state) {
                        State.NORMAL -> State.EDIT
                        else -> state
                    }
                    onLongPress()
                })) {
            Text(text = friend.name,
                    style = MaterialTheme.typography.subtitle1,
                    modifier = Modifier
                            .alpha(if (state == State.NORMAL) 1F else 0.5F)
                            .blur(if (state ==
                                    State.NORMAL) 0.dp else 5.dp))
            when (state) {
                State.EDIT -> {
                    Row {
                        Button(onClick = { state = State.NORMAL },
                                colors = ButtonDefaults.buttonColors(
                                        backgroundColor = MaterialTheme.colors.background,
                                        contentColor = MaterialTheme.colors.onBackground
                                )) {

                            Icon(Icons.Filled.Close,
                                    contentDescription = getString(R.string.cancel))
                        }
                        Button(onClick = { onDeletePrompt(friend) }, colors =
                        ButtonDefaults
                                .buttonColors(backgroundColor = MaterialTheme.colors.error,
                                        contentColor = MaterialTheme.colors.onError)) {
                            Text(getString(R.string.delete))
                        }
                        Button(onClick = { onEditName(friend) }, colors = ButtonDefaults
                                .buttonColors(backgroundColor = MaterialTheme.colors.secondary,
                                        contentColor = MaterialTheme.colors.onSecondary)) {
                            Text(getString(R.string.rename))
                        }
                    }
                }
                State.CONFIRM -> {
                    Row {
                        Button(onClick = { state = State.NORMAL },
                                colors = ButtonDefaults.buttonColors(
                                        backgroundColor = MaterialTheme.colors.background,
                                        contentColor = MaterialTheme.colors.onBackground
                                )) {

                            Icon(Icons.Filled.Close,
                                    contentDescription = getString(R.string.cancel))
                        }
                        Button(onClick = {
                            state = State.CANCEL
                            message = null
                        }, colors =
                        ButtonDefaults
                                .buttonColors(backgroundColor = MaterialTheme.colors.error,
                                        contentColor = MaterialTheme.colors.onError)) {
                            Text(getString(R.string.confirm_alert))
                        }
                        Button(onClick = {
                            /*val context = applicationContext
                            val builder = AlertDialog.Builder(context)
                            builder.setTitle(getString(R.string.add_message))
                            val input = EditText(context)
                            input.setHint(R.string.message_hint)
                            input.inputType =
                                    InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_AUTO_COMPLETE or
                                            InputType.TYPE_TEXT_FLAG_AUTO_CORRECT or
                                            InputType.TYPE_TEXT_FLAG_CAP_SENTENCES or
                                            InputType.TYPE_TEXT_FLAG_MULTI_LINE or
                                            InputType.TYPE_TEXT_VARIATION_SHORT_MESSAGE
                            builder.setView(input)
                            builder.setPositiveButton(
                                    android.R.string.ok) { _: DialogInterface?, _: Int ->
                                message = input.text.toString()
                                state = State.CANCEL
                            }
                            builder.setNegativeButton(
                                    android.R.string.cancel) { _: DialogInterface?, _: Int -> }
                            builder.show() */
                                         friendModel.appendDialogState(Triple(MainViewModel
                                                 .DialogStatus.ADD_MESSAGE_TEXT, friend) {
                                             message = it
                                             state = State.CANCEL
                                         })
                        }, colors = ButtonDefaults
                                .buttonColors(backgroundColor = MaterialTheme.colors.secondary,
                                        contentColor = MaterialTheme.colors.onSecondary)) {
                            Text(getString(R.string.add_message))
                        }
                    }
                }
                State.CANCEL -> {
                    var progress by remember { mutableStateOf(0f) }
                    val animatedProgress by animateFloatAsState(
                            targetValue = progress,
                            animationSpec = ProgressIndicatorDefaults.ProgressAnimationSpec
                    )

                    val delay = object : CountDownTimer(UNDO_TIME, UNDO_TIME / UNDO_INTERVALS) {
                        override fun onTick(p0: Long) {
                            progress = (UNDO_TIME - p0).toFloat() / UNDO_TIME
                        }

                        override fun onFinish() {
                            friendModel.sendAlert(friend.id, message = message)
                        }
                    }

                    LinearProgressIndicator(progress = animatedProgress,
                            modifier = Modifier.clickable {
                                delay.cancel()
                                state = State.NORMAL
                            })


                    delay.start()
                }
                else -> {}
            }

        }
    }

    @ExperimentalFoundationApi
    @Preview
    @Composable
    fun PreviewHome() {
        MaterialTheme(colors = if (isSystemInDarkTheme()) darkColors else lightColors) {
            Home(listOf(Friend("1", "Grace"), Friend("2", "Anita")), {}, {}, {},
                    Triple(MainViewModel.DialogStatus.NONE, null) {}, false)
        }
    }

    public override fun onResume() {
        super.onResume()
        // if Google API isn't available, do this - it's from the docs, should be correct
        if (!checkPlayServices()) return
    }

    private fun checkPlayServices(): Boolean {
        if (GoogleApiAvailability.getInstance()
                        .isGooglePlayServicesAvailable(this) != ConnectionResult.SUCCESS) {
            // check for Google Play Services
            Toast.makeText(this, getString(R.string.no_play_services), Toast.LENGTH_LONG).show()
            GoogleApiAvailability.getInstance().makeGooglePlayServicesAvailable(this)
            return false
        }
        return true
    }

    companion object {
        // time (in milliseconds) that the user has to cancel sending an alert
        private const val UNDO_TIME: Long = 3500
        private const val UNDO_INTERVALS: Long = 10


    }
}